assert on all_reduce_event only if it's not CPU device.

Ritesh1905 opened the pull request:
Summary: For CPU based runs, `all_reduce_event` would be None since this is the result of the `all_reduce_stream.record_event()`, which does not do much other than returning None when device type is CPU.

Test Plan: CI

Differential Revision: D72176406




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 commented:
@albanD Is it true that for CPU, `stream.record_event()` would return None? Just for our education.

kwen2501 reviewed (approved):
The change looks okay to me. Asking @albanD for guidance here.

Ritesh1905 committed:
assert on all_reduce_event only if it's not CPU device. (#150316)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/150316

For CPU based runs, `all_reduce_event` would be None since this is the result of the `all_reduce_stream.record_event()`, which does not do much other than returning None when device type is CPU.

Test Plan: CI

Reviewed By: kwen2501

Differential Revision: D72176406

mori360 reviewed (approved):
The pending check is unstable, https://github.com/pytorch/pytorch/issues/150420

awgu commented on torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py:
```diff
@@ -414,11 +414,14 @@ def post_backward(self, *unused: Any):
         with record_function(self._with_fqn("FSDP::post_backward_reduce")):
-            if self.comm_ctx.reduce_scatter_state is not None:
+            if (
+                self.comm_ctx.reduce_scatter_state is not None
+                and self.comm_ctx.reduce_scatter_state.event is not None
+            ):
                 self.device_handle.current_stream().wait_event(
                     self.comm_ctx.reduce_scatter_state.event
                 )
-                self.comm_ctx.reduce_scatter_state = None
+            self.comm_ctx.reduce_scatter_state = None
```
nit: maybe it's better to keep the original style and do a double if statement like:
```
if self.comm_ctx.reduce_scatter_state is not None:
    if self.comm_ctx.reduce_scatter_state.event is not None:
        self.device_handle.current_stream().wait_event(
            self.comm_ctx.reduce_scatter_state.event
        )
    self.comm_ctx.reduce_scatter_state = None
```
there is semantic meaning to `if self.comm_ctx.reduce_scatter_state is not None:` -- namely that there was a preceding reduce-scatter that we have not waited on
if we ever want to refactor the stashing and waiting logic to be handled in a single API, it would be clearer if we only cleared when we need to
